### PR TITLE
ldc: 1.30.0 -> 1.31.0, bump bootstrap ldc to 1.30.0, bump LLVM to 14

### PR DESCRIPTION
**Pull request by GZGavinZhao**
###### Description of changes

- ldc: 1.30.0 -> 1.31.0, changelog available [here](https://github.com/ldc-developers/ldc/releases/tag/v1.31.0). Also update the bootstrap LDC's version to 1.30.0 and build against LLVM 14 instead of LLVM 11.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by GZGavinZhao** [comment]
Oops, sorry for the hash mismatch. I think I got confused with which bootstrap version I was using. I just double checked all the hashes. Thanks!

**Comment by GZGavinZhao** [comment]
I was not able to run `nixpkgs-review` (using `rev HEAD` or `pr 218905`) because it hangs on the first command: `git -c fetch.prune=false fetch --no-tags --force https://github.com/NixOS/nixpkgs master:refs/nixpkgs-review/0`. I shallow cloned the `nixpkgs` with depth 1. Could this be the issue?

**Comment by GZGavinZhao** [comment]
So it does need a full clone... Successfully ran `nixpkgs-review`.

**Comment by dukc** [comment]
ping @ThomasMader @lionello 

I think Thomas wants to update DMD and LDC in lockstep. However, I just created [another MR] regarding DMD which should make a matter of only a hash update.

**Review by lionello — APPROVED** [review]

**Review by beardhatcode — APPROVED** [review]
Successfully used the `ldc2` compiler from this PR to compile a large software program :+1:

File: pkgs/development/compilers/ldc/bootstrap.nix [hunk 1]
```diff
@@ -1,11 +1,11 @@
...
 callPackage ./binary.nix {
-  version = "1.25.0";
+  version = "1.30.0";
   hashes = {
-    # Get these from `nix-prefetch-url https://github.com/ldc-developers/ldc/releases/download/v1.19.0/ldc2-1.19.0-osx-x86_64.tar.xz` etc..
-    osx-x86_64 = "sha256-6iKnbS+oalLKmyS8qYD/wS21b7+O+VgsWG2iT4PrWPU=";
-    linux-x86_64 = "sha256-sfg47RdlsIpryc3iZvE17OtLweh3Zw6DeuNJYgpuH+o=";
-    linux-aarch64  = "sha256-UDZ43x4flSo+SfsPeE8juZO2Wtk2ZzwySk0ADHnvJBI=";
-    osx-arm64  = "sha256-O/x0vy0wwQFaDc4uWSeMhx+chJKqbQb6e5QNYf+7DCw=";
+    # Get these from `nix store prefetch-file https://github.com/ldc-developers/ldc/releases/download/v1.19.0/ldc2-1.19.0-osx-x86_64.tar.xz` etc..
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/ldc/bootstrap.nix:5]: we should add an update script

**Comment by kvtb** [comment]
it would be nice to set `doCheck = true`, it fails some tests and the current setup -- having `checkPhase` without `doCheck` (which defaults to `false`) -- hides the problem

**Pull request merged**